core/builder/src/main/java/io/quarkus/builder/Multis.java

```diff
@@ -0,0 +1,165 @@
+package io.quarkus.builder;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.quarkus.builder.Multis.TopoMultiList.TopoEntry;
+import io.quarkus.builder.item.MultiBuildItem;
+
+/**
+ * A facility for storing and retrieving {@link MultiBuildItem}s by {@link ItemId}.
+ */
+class Multis {
+    /** A special ordinal to use for entries that are available initially before the {@link Execution} has started */
+    private static final int INITIAL_ORDINAL = -1;
+
+    private final ConcurrentHashMap<ItemId, TopoMultiList<MultiBuildItem>> multis;
```

Comment:
This seems overly complex.  Could it not be a `List<List<MultiBuildItem>>`?  The outer index would be the position relative to the ordinal of the producing step.  The size of the outer list should be known ahead of time because it is bounded by the number of producing steps (which should be a small-ish number).  You could treat a `null` entry as an empty list to avoid constructing useless objects.

Reply:
See 26312da